modules/src/applications/transfer/context.rs

```diff
@@ -0,0 +1,348 @@
+
+    /// Returns the escrow account id for a port and channel combination
+    fn get_channel_escrow_address(
+        &self,
+        port_id: &PortId,
+        channel_id: ChannelId,
+    ) -> Result<<Self as Ics20Reader>::AccountId, Ics20Error> {
+        // https://github.com/cosmos/cosmos-sdk/blob/master/docs/architecture/adr-028-public-key-addresses.md
+        let contents = format!("{}/{}", port_id, channel_id);
+
+        let mut hasher = Sha256::new();
+        hasher.update(VERSION.as_bytes());
+        hasher.update(b"0");
+        hasher.update(contents.as_bytes());
+        let hash: Vec<u8> = hasher.finalize().to_vec().drain(20..).collect();
+
+        String::from_utf8(hex::encode_upper(hash))
+            .expect("hex encoded bytes are not valid UTF8")
+            .parse::<Signer>()
+            .map_err(Ics20Error::signer)?
+            .try_into()
+            .map_err(|_| Ics20Error::parse_account_failure())
+    }
+
+    /// Returns true iff send is enabled.
+    fn is_send_enabled(&self) -> bool;
+
+    /// Returns true iff receive is enabled.
+    fn is_receive_enabled(&self) -> bool;
+
+    /// Returns true iff the store contains a `DenomTrace` entry for the specified `HashedDenom`.
+    fn has_denom_trace(&self, hashed_denom: &HashedDenom) -> bool {
+        self.get_denom_trace(hashed_denom).is_some()
+    }
+
+    /// Get the denom trace associated with the specified hash in the store.
+    fn get_denom_trace(&self, denom_hash: &HashedDenom) -> Option<DenomTrace>;
+}
+
+pub trait BankKeeper {
+    type AccountId: ToString;
+
+    /// This function should enable sending ibc fungible tokens from one account to another
+    fn send_coins(
+        &mut self,
+        from: &Self::AccountId,
+        to: &Self::AccountId,
+        amt: &IbcCoin,
+    ) -> Result<(), Ics20Error>;
+
+    /// This function to enable minting ibc tokens in a module
+    fn mint_coins(&mut self, module: &Self::AccountId, amt: &IbcCoin) -> Result<(), Ics20Error>;
+
+    /// This function should enable burning of minted tokens
+    fn burn_coins(&mut self, module: &Self::AccountId, amt: &IbcCoin) -> Result<(), Ics20Error>;
+
+    /// This function should enable transfer of tokens from the ibc module to an account
+    fn send_coins_from_module_to_account(
+        &mut self,
+        module: &Self::AccountId,
+        to: &Self::AccountId,
+        amt: &IbcCoin,
+    ) -> Result<(), Ics20Error>;
+
+    /// This function should enable transfer of tokens from an account to the ibc module
+    fn send_coins_from_account_to_module(
+        &mut self,
+        from: &Self::AccountId,
+        module: &Self::AccountId,
+        amt: &IbcCoin,
+    ) -> Result<(), Ics20Error>;
+}
+
+pub trait BankReader {
+    type AccountId: ToString + TryFrom<Signer>;
+
+    /// Returns true if the specified account is not allowed to receive funds and false otherwise.
+    fn is_blocked_account(&self, account: &Self::AccountId) -> bool;
```

Comment:
Do we really need to expose `is_blocked_account()` and run the check ourselves in `process_recv_packet()`? The `BankKeeper::send_coins()` function is allowed to fail, so we can leave it to the implementations to check if the receiver address is blocked before transferring the funds (which they should do anyways).

Reply:
Great idea! Removed `is_blocked_account()` -> https://github.com/informalsystems/ibc-rs/pull/1989/commits/2c61fc296fdc6f3258793ed7809a9b46ae922d2f